docs/dev/relay_add_pass.rst

```diff
@@ -0,0 +1,125 @@
+
+Background
+----------
+
+Constant folding involves evaluating expressions in the program that only
+involve constant values, then replacing those expressions with the result
+of evaluating them. The goal of this pass is to frontload all of the
+computations that we can. With that background, we now present an example of
+an ``ExprVisitor``, then an example of an ``ExprMutator``.
+
+Writing an Expression Visitor
+-----------------------------
+
+The constant folding pass makes use of a ``ConstantChecker``, which extends
+``ExprVisitor``. The public interface to this class is a ``Check`` method
+that returns whether the given expression is considered constant.
+In Relay, we define an expression to be constant if it is a
+``ConstantNode`` or it is a ``TupleNode`` with only constant fields.
+
+Internally, the ``ConstantChecker`` uses the ``VisitExpr_`` functions to
+check for constness. Because ``VisitExpr_`` returns nothing, we need to keep
+track of our results in our own data structure. So we use a ``memo_`` field
+to map from nodes to whether they are constant. Additionally, ``memo_``
+serves as a cache for later ``Check`` calls on the same expression. Below are
+the ``VisitExpr_`` definitions in the ``ConstantChecker``.
+
+.. code:: c
+
+    void VisitExpr_(const ConstantNode* n) final {
+      memo_[GetRef<Constant>(n)] = true;
+    }
+
+    void VisitExpr_(const TupleNode* n) final {
+      bool result = true;
+      for (const auto& field : n->fields) {
+        if (!Check(field)) {
+          result = false;
+          break;
+        }
+      }
+      memo_[GetRef<Tuple>(n)] = result;
+    }
+
+Note that we don't need to define a ``VisitExpr_`` function for every type of
+node, because ``ExprVisitor`` defines a default implementation for every
+node. This default implementation simply visits all of its fields that are
+nodes. So we can default to returning false and only define ``VisitExpr_``
+methods for the node types that could possibly change the result to true
+(i.e., the ones listed in our constness condition earlier).
+
+Writing an Expression Mutator
+-----------------------------
+
+The second component of the constant folding pass is the ``ConstantFolder``,
+which extends ``ExprMutator`` and internally uses ``ConstantChecker``. The
+``ConstantFolder`` class is used to implement the ``FoldConstant`` function,
+so it can be registered to become an API endpoint.
+
+In the ``ConstantChecker``, we defined our own ``memo_`` field, but we don't
+need to do the same in ``ConstantFolder``. ``ExprMutator`` already provides a
+memoizer, because we know its ``VisitExpr_`` functions will always take
+expressions and return expressions.
+
+Similarly to ``ExprVisitor``, ``ExprMutator`` also defines default
+implementations for all node types. Each ``VisitExpr_`` method visits all of
+its node fields and creates a new node if any of the resulting fields have
+changed. Otherwise, the original node is returned.
+
```

Comment:
I think the description of a mutator vs visitor should come earlier and be more detailed, since the visitor and mutator are very important base classes that anyone writing a pass will have to interact with.